### ISSUE TITLE
IOCP reactor crash when using UDP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5135]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5135|
|Type|defect|
|Created|2011-06-05 23:50:18Z|

Apparently the attached code causes Twisted to segfault. More details forthcoming if original reporter shows up (`mat^2` on IRC).

Attachments:

* [iocpcrash.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/62d/62d00299f49f77c3e7ef2726b826b3e6871b6e22/cc752979d31efff8484fa98f3c3ce92cf709a06c.txt) (1425 bytes) - added by itamar on 2011-06-05 23:50:37Z - Scripts to demonstrate crash

<details><summary>Searchable metadata</summary>

```
trac-id__5135 5135
type__defect defect
reporter__itamar itamar
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__IOCP_windows IOCP windows
time__1307317818000000 1307317818000000
changetime__1504512539014478 1504512539014478
version__None None
owner__ 
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank commented|
|-|-|

On Server 2008 R2 SP1 64-bit, neither server.py nor client.py crashes for me (tested official Python 2.7.1 32-bit, official Twisted 11.0.0).

Can you provide more details about your OS/Python/Twisted?  Does it always crash?

This *might* be related to [#3970](https://github.com/twisted/twisted/issues/3970)

**Comment by twisted-trac** [comment]
|<img alt="matpow2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|matpow2 commented|
|-|-|

On Windows XP SP3, 32bit, it gives a consistent crash for me using Twisted 11.0.0, 2.6.6. I'll try with Python 2.7.1 soon and see if it makes a difference.

**Comment by twisted-trac** [comment]
|<img alt="matpow2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|matpow2 commented|
|-|-|

Oh, sorry, I should probably also note that it's the *client* that's crashing, not the server. Lowering the number of clients makes it not segfault, it seems.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

I can't reproduce the crash.  I used a clean XP SP3 32bit install in VMWare Workstation, with all of the important updates as of today (2011-06-06).  I installed:

```
python-2.6.6.msi
zope.interface-3.6.3.tar.gz (from source, no optimizations due to lack of compiler)
Twisted-11.0.0.winxp32-py2.6.exe
pywin32-214.win32-py2.6.exe
```

I then tried again after installing `pyOpenSSL-0.11.winxp32-py2.6.msi`, but it still didn't crash.  I also tried starting client.py before server.py, but that didn't crash either.

Note that when Windows Firewall asked, I chose to "Unblock" server.py.

Does your zope.interface have optimizations (and which z.i version)?  Which pywin32 did you install?  Do you have pyOpenSSL installed (which version?)  What other modules do you have loaded in memory?  Are you using any AV software, or any software that injects itself into every process?

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

I tried it with `pywin32-216.win32-py2.6.exe` just now, still no crash.

**Comment by twisted-trac** [comment]
|<img alt="matpow2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|matpow2 commented|
|-|-|

I use zope.interface 3.5.2 with optimizations, pywin32-216. I don't have pyOpenSSL installed. I don't use any firewall or antivirus or anything that would hook into processes.

This is the output of sys.modules: http://paste.pocoo.org/show/KQVOOZfUozfNpDT1PofY/

This is the output with logging turned on: http://paste.pocoo.org/show/yGhiMpN3FyISJYQqwP52/

I'll try and get a debugger on this (need to build with debug first, I suppose).

**Comment by twisted-trac** [comment]
|<img alt="matpow2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|matpow2 commented|
|-|-|

After rebooting, I don't get this issue anymore. I'm not sure whether this ticket should be closed, as I can't reproduce it now.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ivank commented|
|-|-|

This shouldn't be closed; it's probably the same bug as [#3970](https://github.com/twisted/twisted/issues/3970).  This crash *might* be happening only when something is allocated in a certain region of memory.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc set status to `closed`|
|-|-|

Fixed as a side-effect of [ticket:8870]
